snappy revamp

arnetheduck opened the pull request:
This is a more or less complete revamp of the snappy library aiming to:

* clear out a lot of the duplicate code
* remove some of the redundant API
* unify the codebase behind a single, optimized "inner" encoder/decoder
* unify the public API for in-memory and stream
compression/decompression
* improve performance

As such, only the documented API remains backwards-compatible - the rest
has been refactored, moved around and rewritten:

* `import snappy` now exposes only in-memory encoders / decoders
* framed format moved to `snappy` module, `snappy/framing` removed
* faststreams integration moved to `snappy/faststreams`
* minimal `std/streams` integration started in `snappy/streams`

Other changes include:

* up-to-date documentation
* allocation- and exception-free API (uses some amount of stack memory)
* a 2-3x improvement to both compression and decompression performance,
putting the library mostly on par with the C++ implementation (see
README)
* the implementation was heavily inspired by the `C++` and `go`
implementations, but somewhat simplified
* nonetheless, the code uses a significant amount of unsafe code to
work around inefficiencies in the safe subset of Nim

arnetheduck committed:
Snappy revamp

This is a more or less complete revamp of the snappy library aiming to:

* clear out a lot of the duplicate code
* remove some of the redundant API
* unify the codebase behind a single, optimized "inner" encoder/decoder
* unify the public API for in-memory and stream
compression/decompression
* improve performance

As such, only the documented API remains backwards-compatible - the rest
has been refactored, moved around and rewritten:

* `import snappy` now exposes only in-memory encoders / decoders
* framed format moved to `snappy` module, `snappy/framing` removed
* faststreams integration moved to `snappy/faststreams`
* minimal `std/streams` integration started in `snappy/streams`

Other changes include:

* up-to-date documentation
* allocation- and exception-free API (uses some amount of stack memory)
* a 2-3x improvement to both compression and decompression performance,
putting the library mostly on par with the C++ implementation (see
README)
* the implementation was heavily inspired by the `C++` and `go`
implementations, but somewhat simplified
* nonetheless, the code uses a significant amount of unsafe code to
work around inefficiencies in the safe subset of Nim

arnetheduck committed:
test fixes

arnetheduck committed:
fix i386

arnetheduck committed:
cleanups

arnetheduck committed:
flush outputs

arnetheduck committed:
compare encoded streams as well

arnetheduck committed:
try removing runway

arnetheduck committed:
clean up `uint` logic

arnetheduck committed:
fix fuzzing compilation

...but running them seems messy...

arnetheduck committed:
more fuzzing fixes

arnetheduck committed:
tighten max frame data length error conditions

jangko reviewed (changes requested):
The one in `encodeFrame` function of `encoder.nim` looks serious enough.

arnetheduck committed:
allow resuming decompression

* typos
* fix framed minimum input length check
* add option to skip header (for partial reads)
* fix minimum compressed length for framed data with a small last frame

jangko reviewed (approved):
LGTM

arnetheduck committed:
move back to a (mostly) range-checked implementation

with bulk operations in place, the cost of range checks falls
significantly - we can reintroduce them without any significant loss in
performance by carefully ordering operations such that optimizers can
elide most.

* better copy decoding implementation that extends offset so that bulk
copying can be used - 20% faster decoding

arnetheduck committed:
touchups
